azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message_queue.py

```diff
@@ -0,0 +1,142 @@
+        "input" , "c2d" etc.
+        "inputs" and on the attribute "on_c2d_message" for "c2d"
+        :raises ValueError if feature_name is not amongst "input" or "c2d"
+        """
+        enable_complete = Event()
+
+        def callback():
+            enable_complete.set()
+
+        if feature_name == "input":
+            self._transport.enable_input_messages(callback)
+        elif feature_name == "c2d":
+            self._transport.enable_c2d_messages(callback)
+        else:
+            logger.error("Feature name has not been defined.Feature names can be 'input' or 'c2d'")
+            raise ValueError("Feature names can be only among 'input' or 'c2d'")
+
+        enable_complete.wait()
+
+    def _disable_feature(self, feature_name):
+        disable_complete = Event()
+
+        def callback():
+            disable_complete.set()
+
+        if feature_name == "input":
+            self._transport.disable_input_messages(callback)
+        elif feature_name == "c2d":
+            self._transport.disable_c2d_messages(callback)
+        else:
+            logger.error("Feature name has not been defined.Feature names can be 'input' or 'c2d'")
+            raise ValueError("Feature names can be only among 'input' or 'c2d'")
+
+        disable_complete.wait()
+
+    def get_input_message_queue(self, input_name):
+        try:
+            queue = self.input_message_queues[input_name]
+        except KeyError:
+            # Create new MessageQueue for input if it does not yet exist
+            queue = MessageQueue(
+                self._input_message_queue_on_enable, self._input_message_queue_on_disable
+            )
+            self.input_message_queues[input_name] = queue
+
+        if queue.enabled is not True:
+            queue.enable()
+        return queue
+
+    def get_c2d_message_queue(self):
+        if self.c2d_message_queue is None:
+            queue = MessageQueue(
+                self._c2d_message_queue_on_enable, self._c2d_message_queue_on_disable
+            )
+            self.c2d_message_queue = queue
+        else:
+            queue = self.c2d_message_queue
+
+        if queue.enabled is not True:
+            queue.enable()
+        return queue
+
+    def receive_input_message(self, input_name, message_received):
+        try:
+            queue = self.input_message_queues[input_name]
+        except KeyError:
+            logger.error("No input message queue for {}".format(input_name))
+        else:
+            if queue.enabled:
+                queue.put(message_received)
+                logger.info("Input message delivered to {} queue".format(input_name))
+            else:
+                logger.warn("Input message queue for {} is disabled".format(input_name))
+
+    def receive_c2d_message(self, message_received):
+        # No check on if the queue exists - it should.
+        # If for some reason it doesn't, exception will be raised
```

Comment:
>exception will be raised [](start = 41, length = 24)

This shouldn't throw.  This is very possible -- for MQTT at least, subscription state survives a disconnect/reconnect.  So, if you are subscribed, then you reboot your device and reconnect, you're still subscribed even though the caller might never have called any enable functions.